File: sdks/python/tests/_support.py
"""Apoio dos testes: servidor falso (FAKE_API_URL ou iniciado aqui) e leitura do que ele recebeu."""

from __future__ import annotations

import json
import os
import re
import sys
import urllib.request
import uuid
from pathlib import Path
from typing import Any

REPO = Path(__file__).resolve().parents[3]
VECTORS = REPO / "sdks" / "testdata" / "webhook-signature-vectors.json"
UUID_V4 = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")

_url: str | None = None
_server: Any = None
_direct = urllib.request.build_opener(urllib.request.ProxyHandler({}))


def fake_url() -> str:
    """URL base do servidor falso. Sem FAKE_API_URL, sobe um na própria thread de teste."""
    global _url, _server
    if _url:
        return _url
    configured = os.environ.get("FAKE_API_URL")
    if configured:
        _url = configured.rstrip("/")
        return _url
    sys.path.insert(0, str(REPO / "tools" / "sdkgen"))
    import fake_server  # noqa: PLC0415

    _server, _url = fake_server.start_in_thread()
    return _url


def origin() -> str:
    url = fake_url()
    return url[: url.index("/api/v1")]


def new_trace() -> tuple[dict[str, str], str]:
    trace = uuid.uuid4().hex
    return {"X-Fake-Trace": trace}, trace


def fetch_trace(trace: str) -> list[dict[str, Any]]:
    with _direct.open(f"{origin()}/__fake/requests?trace={trace}", timeout=10) as response:
        return json.loads(response.read().decode("utf-8"))


def free_port_url() -> str:
    """Endereço sem ninguém escutando (conexão recusada)."""
    import socket

    with socket.socket() as sock:
        sock.bind(("127.0.0.1", 0))
        port = sock.getsockname()[1]
    return f"http://127.0.0.1:{port}/api/v1"


def load_vectors() -> dict[str, Any]:
    return json.loads(VECTORS.read_text(encoding="utf-8"))

File: sdks/node/test/support.mjs
import { randomUUID } from 'node:crypto';
import { readFileSync } from 'node:fs';

export const UUID_V4 =
    /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;

/** URL do servidor falso (tools/sdkgen/fake_server.py). */
export function fakeUrl() {
    const url = process.env.FAKE_API_URL;
    if (!url) {
        throw new Error(
            'FAKE_API_URL não definido. Rode: python tools/sdkgen/sdkgen.py test',
        );
    }
    return url.replace(/\/+$/, '');
}

/** Opções com o cabeçalho de rastreio e o id para consultar o servidor falso. */
export function newTrace(extra = {}) {
    const trace = randomUUID().replaceAll('-', '');
    return [{ ...extra, headers: { 'X-Fake-Trace': trace } }, trace];
}

export async function fetchTrace(trace) {
    const url = fakeUrl();
    const origin = url.slice(0, url.indexOf('/api/v1'));
    const response = await fetch(`${origin}/__fake/requests?trace=${trace}`);
    return response.json();
}

export async function lastRecord(trace) {
    const records = await fetchTrace(trace);
    if (records.length === 0) {
        throw new Error('o servidor falso não recebeu o pedido');
    }
    return records[records.length - 1];
}

export function loadVectors() {
    return JSON.parse(
        readFileSync(
            new URL(
                '../../testdata/webhook-signature-vectors.json',
                import.meta.url,
            ),
            'utf8',
        ),
    );
}
